File: universal7885-common/SamsungParts/src/com/eurekateam/samsungextras/DeviceSettings.java
/*
 * Copyright (C) 2020 The Xiaomi-SM6250 Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 */

package com.eurekateam.samsungextras;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import androidx.preference.Preference;
import androidx.preference.PreferenceFragment;
import androidx.preference.PreferenceManager;
import androidx.preference.SwitchPreference;

import com.eurekateam.samsungextras.battery.BatteryActivity;
import com.eurekateam.samsungextras.flashlight.FlashLightActivity;
import com.eurekateam.samsungextras.speaker.ClearSpeakerActivity;
import com.eurekateam.samsungextras.utils.FileUtilsWrapper;
import com.eurekateam.samsungextras.utils.SystemProperties;

public class DeviceSettings extends PreferenceFragment implements
        Preference.OnPreferenceChangeListener {
    public static final String PREF_KEY_FPS_INFO = "fps_info";
    private static final String PREF_CLEAR_SPEAKER = "clear_speaker_settings";
    private static final String PREF_FLASHLIGHT = "flashlight_settings";
    public static final String PREF_GPUEXYNOS = "gpuexynos_settings";
    private static final String PREF_SELINUX = "selinux_settings";
    public static final String PREF_BATTERY = "battery_settings";

    public DeviceSettings() {
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        System.loadLibrary("libnativebridges");
        setPreferencesFromResource(R.xml.preferences_samsung_parts, rootKey);
        Context mContext = this.getContext();
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(mContext);
        Preference mClearSpeakerPref = findPreference(PREF_CLEAR_SPEAKER);
        assert mClearSpeakerPref != null;
        mClearSpeakerPref.setOnPreferenceClickListener(preference -> {
            Intent intent = new Intent(getActivity().getApplicationContext(), ClearSpeakerActivity.class);
            startActivity(intent);
            return true;
        });

        SwitchPreference mFpsInfo = findPreference(PREF_KEY_FPS_INFO);
        assert mFpsInfo != null;
        mFpsInfo.setChecked(prefs.getBoolean(PREF_KEY_FPS_INFO, false));
        mFpsInfo.setOnPreferenceChangeListener(this);

        SwitchPreference mGPUExynos = findPreference(PREF_GPUEXYNOS);
        assert mGPUExynos != null;
        if (FileUtilsWrapper.isFileReadable(GlobalConstants.TMU_SYSFS)) {
            Log.d(GlobalConstants.TAG, "onCreatePreferences: "
                    + (GlobalConstants.TMU_SYSFS) + " readable, value " +
                    FileUtilsWrapper.readOneLine(GlobalConstants.TMU_SYSFS));
            mGPUExynos.setChecked(
                    FileUtilsWrapper.readOneLine(GlobalConstants.TMU_SYSFS).equals("1"));
        }else {
            Log.w(GlobalConstants.TAG, "onCreatePreferences: "
                    + (GlobalConstants.TMU_SYSFS) + " not readable");
            mGPUExynos.setEnabled(false);
        }
        mGPUExynos.setOnPreferenceChangeListener(this);

        Preference mSELinux = findPreference(PREF_SELINUX);
        assert mSELinux != null;
        mSELinux.setOnPreferenceClickListener(preference -> {
            boolean selinux_enforcing = SystemProperties.getenforce();
            Toast.makeText(getContext(), selinux_enforcing ?
                            "SELinux is in enforcing state." :
                            "SELinux is in permissive state.",
                    Toast.LENGTH_SHORT).show();
            return true;
        });
        Preference mFlashLight = findPreference(PREF_FLASHLIGHT);
        mFlashLight.setOnPreferenceClickListener(preference -> {
            Intent intent = new Intent(getActivity().getApplicationContext(), FlashLightActivity.class);
            startActivity(intent);
            return true;
        });
        Preference mFastCharge = findPreference(PREF_BATTERY);
        assert mFastCharge != null;
        mFastCharge.setOnPreferenceClickListener(preference -> {
            Intent intent = new Intent(getActivity().getApplicationContext(), BatteryActivity.class);
            startActivity(intent);
            return true;
        });
    }

    @Override
    public boolean onPreferenceChange(Preference preference, Object value) {
        final String key = preference.getKey();
        switch (key) {
            case PREF_KEY_FPS_INFO:
                boolean fps_enabled = (Boolean) value;
                Intent fpsinfo = new Intent(this.getContext(), FPSInfoService.class);
                if (fps_enabled) {
                    this.getContext().startService(fpsinfo);
                } else {
                    this.getContext().stopService(fpsinfo);
                }
                break;
            case PREF_GPUEXYNOS:
                boolean gpu_enabled = (Boolean) value;
                Log.d(GlobalConstants.TAG, "onPreferenceChange: Writing " +
                        gpu_enabled + " to " + GlobalConstants.TMU_SYSFS);
                FileUtilsWrapper.writeLine(GlobalConstants.TMU_SYSFS, gpu_enabled ? "1" : "0");
                Toast.makeText(getContext(), gpu_enabled ? "GPU Throttling is now enabled."
                        : "GPU Throttling is now disabled.",
                        Toast.LENGTH_SHORT).show();
            default:
                break;
        }
        return true;
    }
}

File: universal7885-common/apps/SamsungParts/src/com/eurekateam/samsungextras/DeviceSettings.kt
/*
 * Copyright (C) 2022 Eureka Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License
 */
package com.eurekateam.samsungextras

import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import androidx.preference.SwitchPreference
import com.eurekateam.samsungextras.battery.BatteryActivity
import com.eurekateam.samsungextras.flashlight.FlashLightActivity
import com.eurekateam.samsungextras.fps.FPSInfoService
import com.eurekateam.samsungextras.interfaces.Display
import com.eurekateam.samsungextras.speaker.ClearSpeakerActivity
import com.eurekateam.samsungextras.smartcharge.SmartChargeActivity

class DeviceSettings : PreferenceFragmentCompat(), Preference.OnPreferenceChangeListener {

    private lateinit var mPrefs: SharedPreferences
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        System.loadLibrary("StorageHelper")
        System.loadLibrary("SwapCallback")
        setPreferencesFromResource(R.xml.preferences_samsung_parts, rootKey)
        mPrefs = PreferenceManager.getDefaultSharedPreferences(requireContext())
        val mClearSpeakerPref = findPreference<Preference>(PREF_CLEAR_SPEAKER)!!
        mClearSpeakerPref.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                val intent = Intent(requireActivity().applicationContext, ClearSpeakerActivity::class.java)
                startActivity(intent)
                true
            }
        val mFpsInfo = findPreference<SwitchPreference>(PREF_KEY_FPS_INFO)!!
        mFpsInfo.isChecked = mPrefs.getBoolean(PREF_KEY_FPS_INFO, false)
        mFpsInfo.onPreferenceChangeListener = this
        val mDT2W = findPreference<SwitchPreference>(PREF_DOUBLE_TAP)!!
        mDT2W.onPreferenceChangeListener = this
        mDT2W.isEnabled = !(Build.PRODUCT.contains("a10") || Build.PRODUCT.contains("a20e"))
        mDT2W.isChecked = mPrefs.getBoolean(PREF_DOUBLE_TAP, false)
        val mGloveMode = findPreference<SwitchPreference>(PREF_GLOVE_MODE)!!
        mGloveMode.onPreferenceChangeListener = this
        mGloveMode.isEnabled = !Build.PRODUCT.contains("a10")
        mGloveMode.isChecked = mPrefs.getBoolean(PREF_GLOVE_MODE, false)
        val mFlashLight = findPreference<Preference>(PREF_FLASHLIGHT)!!
        mFlashLight.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                val intent = Intent(requireActivity().applicationContext, FlashLightActivity::class.java)
                startActivity(intent)
                true
            }
        val mFastCharge = findPreference<Preference>(PREF_BATTERY)!!
        mFastCharge.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                val intent = Intent(requireActivity().applicationContext, BatteryActivity::class.java)
                startActivity(intent)
                true
            }
        val mSmartCharge = findPreference<Preference>(PREF_SMARTCHARGE)!!
        mSmartCharge.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                val intent = Intent(requireActivity().applicationContext, SmartChargeActivity::class.java)
                startActivity(intent)
                true
            }
    }

    override fun onPreferenceChange(preference: Preference, value: Any): Boolean {
        val mDisplay = Display()
        when (preference.key) {
            PREF_KEY_FPS_INFO -> {
                val mEnabled = value as Boolean
                val mFPSService = Intent(this.context, FPSInfoService::class.java)
                if (mEnabled) {
                    requireContext().startService(mFPSService)
                } else {
                    requireContext().stopService(mFPSService)
                }
                mPrefs.edit().putBoolean(PREF_KEY_FPS_INFO, mEnabled).apply()
            }
            PREF_DOUBLE_TAP -> {
                mDisplay.DT2W = value as Boolean
                mPrefs.edit().putBoolean(PREF_DOUBLE_TAP, value).apply()
            }
            PREF_GLOVE_MODE -> {
                mDisplay.GloveMode = value as Boolean
                mPrefs.edit().putBoolean(PREF_GLOVE_MODE, value).apply()
            }
            else -> {
                return false
            }
        }
        return true
    }

    companion object {
        const val PREF_KEY_FPS_INFO = "fps_info"
        private const val PREF_CLEAR_SPEAKER = "clear_speaker_settings"
        private const val PREF_FLASHLIGHT = "flashlight_settings"
        const val PREF_DOUBLE_TAP = "dt2w_settings"
        const val PREF_GLOVE_MODE = "glove_mode_settings"
        const val PREF_BATTERY = "battery_settings"
        const val PREF_SMARTCHARGE = "smartcharge_settings"
    }
}
